src/nvim/statusline.c

```diff
@@ -653,8 +653,9 @@ void win_redr_custom(win_T *wp, bool draw_winbar, bool draw_ruler)
 /// @param tabtab  Tab clicks definition (can be NULL).
 ///
 /// @return  The final width of the statusline
-int build_stl_str_hl(win_T *wp, char *out, size_t outlen, char *fmt, int use_sandbox, int fillchar,
-                     int maxwidth, stl_hlrec_t **hltab, StlClickRecord **tabtab)
+int build_stl_str_hl(win_T *wp, char *out, size_t outlen, char *fmt, char *opt_name,
+                     int use_sandbox, int fillchar, int maxwidth, int *truncate,
+                     stl_hlrec_t **hltab, StlClickRecord **tabtab)
```

Comment:
I changed this function signature before realizing it was an API function. It's not strictly necessary I think. I could find a different solution if preferred but passing in the option name that the `fmt` format string belongs to allows us to remove some code duplication. Rather than checking for `did_emsg` around each `build_stl_str_hl()` call we can do it in the function and reset the option there.

I am returning the truncated width to make sure that the returned string fits in the number column (up to a maximum of 20). Made sense to me to dynamically make the number column wider when necessary but if we cannot change this function signature(API contract?) a different approach would be needed. Making a best guess of the format string width in `'number_width()` would seems to be the best next option to me. But this could result in unnecessary truncation if the format string width is not proportionate to the line number width.

Reply:
this is not directly exposed as an API function.

Reply:
I see, I take it the change is fine then? I did look around on github, seems lualine uses it for [testing](https://github.com/nvim-lualine/lualine.nvim/blob/8a9c97d111142216c53f83d69030fd9bfe753bb7/tests/statusline.lua#L94), guess they will have to adapt.

Reply:
> seems lualine uses it for [testing](https://github.com/nvim-lualine/lualine.nvim/blob/8a9c97d111142216c53f83d69030fd9bfe753bb7/tests/statusline.lua#L94), guess they will have to adapt.

There is no API contract for FFI consumers, yet. That is 100% unsupported.